pkg/util/crypt/crypt_dev_test.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package crypt
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/sylabs/singularity/internal/pkg/test"
+	"github.com/sylabs/singularity/internal/pkg/util/fs"
+	"github.com/sylabs/singularity/internal/pkg/util/fs/squashfs"
+)
+
+func TestEncrypt(t *testing.T) {
+	test.EnsurePrivilege(t)
+	defer test.ResetPrivilege(t)
+
+	dev := &Device{}
+
+	emptyFile, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatalf("failed to create temporary file: %s", err)
+	}
+	emptyFile.Close()
+	defer os.Remove(emptyFile.Name())
+
+	// Create a dummy squashfs file
+	dummyDir, err := ioutil.TempDir("", "dummy-fs-")
+	if err != nil {
+		t.Fatalf("failed to create temporary directory: %s", err)
+	}
+	defer os.RemoveAll(dummyDir)
+	dummyRootDir := filepath.Join(dummyDir, "root")
+	err = os.MkdirAll(dummyRootDir, 0755)
+	if err != nil {
+		t.Fatalf("failed to create %s: %s", dummyRootDir, err)
+	}
+	dummyRootFile := filepath.Join(dummyRootDir, "EMPTYFILE")
+	err = fs.Touch(dummyRootFile)
+	if err != nil {
+		t.Fatalf("failed to create dummy file %s: %s", dummyRootFile, err)
+	}
+	squashfsBin, err := squashfs.GetPath()
+	if err != nil {
+		t.Fatalf("failed to get path to squashfs binary: %s", err)
+	}
+	tempTargetFile, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatalf("failed to create temporary file: %s", err)
+	}
+	tempTargetFile.Close()
+	defer os.Remove(tempTargetFile.Name())
+	squashfsArgs := []string{dummyDir, tempTargetFile.Name(), "-noappend"}
+	cmd := exec.Command(squashfsBin, squashfsArgs...)
+	err = cmd.Run()
+	if err != nil {
+		t.Fatalf("failed to create squashfs file: %s", err)
+	}
+
+	tests := []struct {
+		name        string
+		path        string
+		key         []byte
+		skipCleanup bool
+		shallPass   bool
+	}{
+		{
+			name:      "empty path",
+			path:      "",
+			key:       []byte("dummyKey"),
+			shallPass: false,
+		},
+		{
+			name:      "empty file",
+			path:      emptyFile.Name(),
+			key:       []byte("dummyKey"),
+			shallPass: false,
+		},
+		{
+			name:      "valid file",
+			path:      tempTargetFile.Name(),
+			key:       []byte("dummyKey"),
+			shallPass: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			devPath, err := dev.EncryptFilesystem(tt.path, tt.key)
+			needCleanup := true
+			if tt.shallPass && err != nil {
+				// cryptsetup is currently creating issues with our CI so
+				// if it is only that, we assume it is fine until we can precisely
+				// figure out why it is not working.
+				if !strings.Contains(err.Error(), "--luks2-metadata-size: unknown option") &&
```

Comment:
The CI issues might be related to the version of cryptsetup present, as we were discussing in the meeting yesterday.

Reply:
https://github.com/sylabs/singularity/issues/3983